odoo/models.py

```diff
@@ -4506,13 +4517,104 @@ def _parent_store_update(self):
+        '''
+
+        IMD = self.env['ir.model.data']
+
+        # This is SQL method of finding matching records (to improve the inefficient code below)
+        # TODO: self._module used below is not correct
+        # TODO: verify company restrictions and security rights (the ORM methods takes care of this)
+
+        key_fields = ', '.join(search_fields)
+        prefixed_key_fields = ', '.join([f'r.{field}' for field in search_fields])
+        key_data = [tuple(item['values'][field] for field in search_fields) for item in data_list]
+
+        sql = f"""
+            CREATE TEMP TABLE temp_{self._table}(
+                {', '.join([f'{f} {self._fields[f].column_type[1]}' for f in search_fields])}
+            );
+            INSERT INTO temp_{self._table}
+            VALUES {','.join(['%s' for d in key_data])};
+            WITH rec_count AS (
+                SELECT
+                    ({key_fields}) AS val_tup,
+                    count(*) as cnt
+                FROM
+                    {self._table}
+                {f'WHERE company_id IS NULL or company_id = {self.env.company.id} ' if 'company_id' in self._fields else None}
+                GROUP BY val_tup
+            ), tmp_data AS (
+                SELECT
+                    DISTINCT ({key_fields}) AS val_tup
+                FROM temp_{self._table}
+            )
+            SELECT
+                {prefixed_key_fields},
+                r.id,
+                i.module || '.' || i.name as external_id,
+                c.cnt
+            FROM {self._table} r
+            INNER JOIN tmp_data t ON ({prefixed_key_fields}) = t.val_tup
+            LEFT JOIN rec_count c ON ({prefixed_key_fields}) = c.val_tup
+            LEFT JOIN ir_model_data i ON r.id = i.res_id AND i.model = '{self._name}'
+            {f'WHERE r.company_id IS NULL or company_id = {self.env.company.id} ' if 'company_id' in self._fields else None}
+        """
+        self.env.cr.execute(sql, tuple(key_data))
+        res = self.env.cr.fetchall()
+        res_dict = {row[:len(search_fields)]: row[len(search_fields):] for row in res}
+        self.env.cr.execute(f'DROP TABLE temp_{self._table}')
+
+        for item in data_list:
+            if not item['xml_id']:
+                key = tuple(item['values'][field] or None for field in search_fields)
+                existing_record_info = res_dict.get(key, (0, None, 0))
+
+                if existing_record_info[2] == 1 and existing_record_info[1]: # a record and there is an xml_id
+                    item['xml_id'] = existing_record_info[1]
+                elif existing_record_info[2] == 1 and existing_record_info[0]: # a record with no xml_id
+                    item['xml_id'] = IMD.create({
+                        'model': self._name,
+                        'res_id': existing_record_info[0],
+                        'name': f'__FOUND_{self._table}_{existing_record_info[0]}',
+                        'module': self._module,
```

Comment:
Attaching imported or exported records to the module of their model is a very bad idea, as it leads to the records being considered "master" data, and automatically deleted (as outdated) on module upgrade.

Reply:
using `__ensure_xml_id`